File: 0399-evaluate-division/0399-evaluate-division.py
from collections import defaultdict

class Solution:
    def calcEquation(self, equations, values, queries):
        graph = defaultdict(dict)
        results = []

        # Build the graph from equations
        for (x, y), val in zip(equations, values):
            graph[x][y] = val
            graph[y][x] = 1.0 / val

        # Perform each query via DFS
        for x, y in queries:
            results.append(self.dfs(x, y, graph, set()))

        return results

    def dfs(self, x, y, graph, visited):
        # If x is not in graph, no path
        if x not in graph:
            return -1.0

        # If we've reached the destination, return 1.0
        if x == y:
            return 1.0

        visited.add(x)

        for neighbor in graph[x]:
            if neighbor not in visited:
                val = self.dfs(neighbor, y, graph, visited)
                if val != -1.0:
                    return graph[x][neighbor] * val

        return -1.0

File: 0399-evaluate-division/0399-evaluate-division.go
type Graph map[string]map[string]float64

func (g Graph) AddEdge(from string, to string, value float64) {
    if _, ok := g[from]; !ok {
        g[from] = make(map[string]float64)
    }
    g[from][to] = value
}

func calcEquation(equations [][]string, values []float64, queries [][]string) []float64 {
    g := make(Graph)
    for i, e := range equations {
        g.AddEdge(e[0], e[1], values[i])
        g.AddEdge(e[1], e[0], 1/values[i])
    }
    res := make([]float64, len(queries))
    for i, q := range queries {
        res[i] = g.DFS(q[0], q[1], map[string]bool{})
    }
    return res
}

func (g Graph) DFS(from string, to string, visited map[string]bool) float64 {
    if _, ok := g[from]; !ok {
        return -1.0
    }
    if from == to {
        return 1.0
    }
    visited[from] = true
    for neighbor, value := range g[from] {
        if visited[neighbor] {
            continue
        }
        temp := g.DFS(neighbor, to, visited)
        if temp > 0 {
            return temp * value
        }
    }
    return -1.0
}
